service/src/parentchain_block_syncer.rs

```diff
@@ -0,0 +1,154 @@
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+use crate::enclave_account;
+use itp_api_client_extensions::{AccountApi, ChainApi};
+use itp_enclave_api::{enclave_base::EnclaveBase, sidechain::Sidechain, EnclaveResult};
+use itp_types::SignedBlock;
+use log::{error, trace};
+use my_node_runtime::{BlockNumber, Header};
+use sp_core::sr25519;
+use sp_runtime::{generic::SignedBlock as SignedBlockT, traits::Block};
+use std::cmp::min;
+use substrate_api_client::{rpc::WsRpcClient, Api, ApiResult};
+
+const BLOCK_SYNC_BATCH_SIZE: u32 = 1000;
+
+pub trait NodeApiFacade {
+	fn get_last_finalized_block(&self) -> SignedBlock;
+	fn tee_nonce(&self) -> u32;
+	fn get_blocks(&self, from: u32, to: u32) -> ApiResult<Vec<SignedBlock>>;
+	fn sync_parentchain<PB: Block>(
+		&self,
+		blocks: &[SignedBlockT<PB>],
+		nonce: u32,
+	) -> EnclaveResult<()>;
+}
+
+impl<EnclaveApi: EnclaveBase + Sidechain> NodeApiFacade for NodeApiDelegate<EnclaveApi> {
+	fn get_last_finalized_block(&self) -> SignedBlock {
+		self.substrate_api.last_finalized_block().unwrap().unwrap()
+	}
+
+	fn tee_nonce(&self) -> u32 {
+		let tee_accountid = enclave_account(&self.enclave_api);
+		self.substrate_api.get_nonce_of(&tee_accountid).unwrap()
+	}
+
+	fn get_blocks(&self, from: u32, to: u32) -> ApiResult<Vec<SignedBlock>> {
+		self.substrate_api.get_blocks(from, to)
+	}
+
+	fn sync_parentchain<PB: Block>(
+		&self,
+		blocks: &[SignedBlockT<PB>],
+		nonce: u32,
+	) -> EnclaveResult<()> {
+		self.enclave_api.sync_parentchain(blocks, nonce)
+	}
+}
+
+pub struct NodeApiDelegate<EnclaveApi: EnclaveBase + Sidechain> {
+	substrate_api: Api<sr25519::Pair, WsRpcClient>,
+	enclave_api: EnclaveApi,
+}
+
+impl<EnclaveApi: EnclaveBase + Sidechain> NodeApiDelegate<EnclaveApi> {
+	pub(crate) fn new(
+		substrate_api: Api<sr25519::Pair, WsRpcClient>,
+		enclave_api: EnclaveApi,
+	) -> Self {
+		NodeApiDelegate { substrate_api, enclave_api }
+	}
+}
+
+pub(crate) struct ParentchainBlockSyncer<NodeApi: NodeApiFacade> {
+	node_api: NodeApi,
+}
+
+impl<NodeApi> ParentchainBlockSyncer<NodeApi>
+where
+	NodeApi: NodeApiFacade,
+{
+	pub fn new(node_api: NodeApi) -> Self {
+		ParentchainBlockSyncer { node_api }
+	}
+
+	/// Fetches the amount of blocks to sync from the parentchain and feeds them to the enclave.
+	/// Returns the latest synced block Header.
+	///
+	pub fn sync_parentchain(&self, last_synced_header: Header) -> Header {
+		trace!("Getting current head");
+		let curr_block: SignedBlock = self.node_api.get_last_finalized_block();
+		let curr_block_number = curr_block.block.header.number;
+
+		let mut until_synced_header = last_synced_header;
+		loop {
+			let block_chunk_to_sync =
+				self.get_block_chunk_to_sync(until_synced_header.number, curr_block_number);
+			println!("[+] Found {} block(s) to sync", block_chunk_to_sync.len());
```

Comment:
I would simply replace this part with the below call.

```suggestion
self.node_api
   .get_blocks(
	last_synced_head_number + 1,
	min(last_synced_head_number + BLOCK_SYNC_BATCH_SIZE, curr_block_number),
	)
	.unwrap()
```